src/main/scala/com/linagora/openpaas/gatling/core/authentication/pkce/PKCESteps.scala

```diff
@@ -69,10 +69,25 @@ object PKCESteps {
       .check(status.is(200),
         jsonPath("$.access_token").find.saveAs("access_token"),
         jsonPath("$.refresh_token").find.saveAs("refresh_token"),
-        jsonPath("$.id_token").find.saveAs("id_token")
+        jsonPath("$.id_token").find.saveAs("id_token"),
+        jsonPath("$.expires_in").find.saveAs("expires_in")
       )
 
-  def renewAccessToken: HttpRequestBuilder =
+  def renewAccessToken: ChainBuilder =
+    doIf(session => {
+      val expiresInInSeconds: Int = session("expires_in").validate[Int].toOption.getOrElse(3601)
+      val tokenAcquisitionTime = session("token_acquisition_time").validate[Long].toOption
+      val lastRenewTime: Option[Long] = session("last_renew").validate[Long].toOption
+      lastRenewTime.orElse(tokenAcquisitionTime) match {
+        case None => false //token never acquired should not happen
+        case Some(last) if System.nanoTime() >= (last + expiresInInSeconds * 1000) => true
```

Comment:
nano is... nano so you should multiply by more than 1000 no?

Reply:
I was sure to type 'currentTimeMillis'.... good catch

Reply:
`currentTimeMillis` seems to not be monotonic: https://stackoverflow.com/questions/2978598/will-system-currenttimemillis-always-return-a-value-previous-calls

so maybe sticking with `nanoTime` is better